cint/ginaccint.1
.TH ginaccint 1 "January, 2000" "GiNaC"
.SH NAME
GiNaC-cint \- An interactive interface for GiNaC based on the Cint C/C++ interpreter
.SH SYNPOSIS
.B ginaccint
.SH DESCRIPTION
.B ginaccint
is an interactive frontend for the GiNaC symbolic computation
framework.  It is a tool that lets you write interactive programs in
C++ that directly make use of GiNaC's classes and thus a more complete
replacement for traditional interactive computer algebra systems than
\fBginsh\fP(1) is.  Programs may be composed as scripts and later compiled
with the native compiler and linked into the system.
.SH USAGE
.SS INPUT FORMAT
After startup, ginsh displays a prompt signifying that it is ready to
accept your input. All C++ statements are valid as input, extended by
GiNaC's numeric or symbolic expressions.  E.g.
.BR fibonacci(24)/1104; 
returns a GiNaC object of class
.BR ex,
, which in this case represents the numeric 42.  Symbols are declared by 
statements as
.nf 
GiNaC> symbol x("x"), y("y"), z;
.fi
which defines two named symbols and an anonymous one for later usage.
All GiNaC methods and functions are available as they would be typed
in C++.  It is not necessary to explicitly invoke a print command as
the last expression is automatically printed:
.nf
GiNaC> pow(x+y,4).expand();
out2 = x^4+4*x^3*y+6*x^2*y^2+4*x*y^3+y^4
.fi
Statements are generally closed by either when a closing brace 
.RB ( } )
matches the first opening brace
.RB ( { ) 
or a semicolon
.RB ( ; )
is encountered.


.SS FUNCTION DEFINITIONS
GiNaC-cint must be put into a special mode in order to define a
function.  This is done with the command
.RB .function .
After that any function definition in valid C++ syntax may be typed
in.  It becomes immediatly available for usage.

.SH EXAMPLES
.nf
GiNaC> symbol x("x"), y("y"), z("z");
GiNaC> ex a = pow(x,2)-x-2;
GiNaC> ex b = pow(x+1,2);
GiNaC> ex s = a/b;
GiNaC> s.diff(x);
out1   -2*(1+x)^(-3)*(-2-x+x^2)+(-1+2*x)*(1+x)^(-2)
GiNaC> s.normal();
out2   (-2+x)*(1+x)^(-1)
GiNaC> .function
next expression can be a function definition
GiNaC> ex EulerNumber(unsigned n)
     > {
     >     symbol x;
     >     ex generator = pow(cosh(x),-1);
     >     return generator.diff(x,n).subs(x==0);
     > }
creating file /tmp/ginac26197caa
GiNaC> EulerNumber(42);
out3   -10364622733519612119397957304745185976310201
GiNaC> quit;
.fi

.SH BUGS
Cint accepts most of K&R and ANSI C/C++ language construct but not
perfect.  In fact, Cint is not aimed to be a 100% ANSI/ISO compliant
C/C++ language processor.  It rather is a portable script language
environment which is close enough to the standard C++.  See the file 
.BR limitati.txt
in your Cint distribution.  Please take the time to track down any bug
you encounter as far as possible and contact Masaharu Goto
<MXJ02154@niftyserve.or.jp> for Cint-related bugs and
<ginac-bugs@ginac.de> for any bugs in the GiNaC engine.

.SH AUTHOR
.TP
The GiNaC Group:
.br
Christian Bauer <Christian.Bauer@uni-mainz.de>
.br
Alexander Frink <Alexander.Frink@uni-mainz.de>
.br
Richard Kreckel <Richard.Kreckel@uni-mainz.de>
.SH SEE ALSO
GiNaC Tutorial \- An open framework for symbolic computation within the
C++ programming language
.PP
CLN \- A Class Library for Numbers, Bruno Haible
.PP
\fBginsh\fP(1)
.SH COPYRIGHT
.SS CINT COPYRIGHT
Copyright \(co of Cint and associated tools are owned by Agilent
Technologies Japan Company and the author.  Acknowledgement to the
author by e-mail is recommended at installation.  Source code, binary
executable or library of Cint and associated tools can be used,
modified and distributed free of charge for any purpose provided that
the copyright notice appear in all copies and that both that copyright
notice and this permission notice appear in documentation.
Registration is requested, at this moment, for commercial use.  Send
e-mail to the author <MXJ02154@niftyserve.or.jp>.  The registration is
free.
.SS GINAC COPYRIGHT
Copyright \(co 1999-2000 Johannes Gutenberg Universit\(:at Mainz, Germany

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


NEWS
This file records noteworthy changes.

0.7.0 ()
* Requires CLN 1.1 now.  Class numeric doesn't use an indirect pointer to the
  actual representation any more.  This is a speedup.
* mul::expand() was reengineered to not allocate temporary excess memory.
* Non-integer powers of a symbol are treated as constants by (l)degree() and
  coeff(). Using these functions on an expression containing such powers used
  to fail with an internal error message. The side-effect is that collect()
  can be used on expressions which are not polynomials.
* Added a man page for the ginac-config script.

0.6.4 (10 August 2000)
* Complete revamp of methods in class matrix.  Some redundant (and poor)
  implementations of elimination schemes were thrown out.  The code is now
  highly orthogonal, more flexible and much more efficient.
* Some long standing and quite nasty bugs were discovered and fixed in the
  following functions: add::normal(), heur_gcd(), sr_gcd() and Order_eval().

0.6.3 (25 July 2000)
* Derivatives are now assembled in a slightly different manner (i.e. they
  might 'look' different on first sight).  Under certain circumstances this
  can result in a dramatic speedup because it gives hashing a better chance,
  especially when computing higher derivatives.
* Some series expansions of built-in functions have been reengineered.
* The algorithm for computing determinants can be chosen by the user.  See
  ginac/flags.h and ginac/matrix.h.
* The Dilogarithm (Li2) now has floating point evaluation, derivative and a
  proper series expansion.
* Namespace 'std' cleanly disentangled, as demanded by ISO/EIC 14882-1998(E).
* Some minor bugfixes, one major lsolve()-bugfix and documentation updates.

0.6.2 (21 June 2000)
* ginaccint.bin is now launched by a binary program instead of by a scripts.
  This allows us to write #!-scripts.  A small test suite for GiNaC-cint was
  added.
* Several minor bugfixes.

0.6.1 (18 May 2000)
* Cleanup in the interface to Cint.  The required version is now Cint 5.14.38.
* Several bugfixes in target install.

0.6.0 (11 May 2000)
* IMPORTANT: Several interface changes make programs written with GiNaC 
  much clearer but break compatibility with older versions:
  - f(x).series(x,p[,o]) -> f(x).series(x==p,o)
  - series(f(x),x,p[,o]) -> series(f(x),x==p,o)
  - gamma() -> tgamma()  (The true Gamma function, there is now also
    log(tgamma()), called lgamma(), in accord with ISO/IEC 9899:1999.)
  - EulerGamma -> Euler
* #include'ing ginac.h defines the preprocessor symbols GINACLIB_MAJOR_VERSION,
  GINACLIB_MINOR_VERSION, and GINACLIB_MICRO_VERSION with the respective GiNaC
  library version numbers.
* Expressions can be constructed from strings like this:
    ex e("2*x+y", lst(x, y));
* ex::to_rational() provides a way to extend the domain of functions like
  gcd() and divide() that only work on polynomials or rational functions (the
  good old ex::subs() method reverses this process)
* Calling diff() on a function that has no derivative defined returns the
  inert derivative function "Derivative".
* Several new timings in the check target.  Some of them may be rather rude 
  at your machine, feel free to interrupt them.

0.5.4 (15 March 2000)
* Some algorithms in class matrix (notably determinant) were replaced by
  less brain-dead ones and should now have much better performance.
* Checks were completely reorganized and split up into three parts:
  a) exams (small regression tests with predefined input)
  b) checks (lenghty coherence checks with random input)
  c) timings (for coherence and crude benchmarking)
* Behaviour of .evalf() was changed: it doesn't .evalf() any exponents.
* Expanded expressions now remember they are expanded to prevent
  superfluous expansions.
* Small bugfixes and improvements in the series expansion.

0.5.3 (23 February 2000)
* A more flexible scheme for registering functions was implemented,
  allowing for remembering, too.
* Some Bugfixes.

0.5.2 (16 February 2000)
* Mainly fixes a bug in the packaging of release 0.5.1.

0.5.1 (14 February 2000)
* Fixes a small number of bugs.

0.5.0 (7 February 2000)
* Expressions can be written ("archived") to files and read therefrom.
* Addition of GiNaC-cint, which lets you write complete programs in
  an interactive shell-like manner in your favoured programming
  language (i.e. C++).

0.4.1 (13 December 1999)
* Series Expansion of Gamma function and some other trigonometric
  functions at their poles works now.
* Many more evaluations of special functions at points where
  exact results exist.
* info_flags::rational doesn't return true for complex extensions
  any more---use info_flags::crational for the old behaviour.
  info_flags::integer and -::cinteger work similarly, the same
  holds for types like info_flags::rational_polynomial.

0.4.0 (26 November 1999)
* First public release.


cint/ginaccint.bin.cpp
#include "G__ci.h"   /* G__atpause is defined in G__ci.h */
#include <iostream>
#include <fstream>
#include <string>
#include <stdio.h>
#include <stdlib.h>
#include "ginac/ginac.h"
#include "config.h"
#include <list>

extern "C" G__value G__exec_tempfile G__P((char *file));
extern "C" void G__store_undo_position(void);

#ifdef OBSCURE_CINT_HACK

#include <strstream>

template<class T>
string ToString(const T & t)
{
    char buf[256];
    ostrstream(buf,sizeof(buf)) << t << ends;
    return buf;
}

basic * ex::last_created_or_assigned_bp=0;
basic * ex::dummy_bp=0;
long ex::last_created_or_assigned_exp=0;

#endif // def OBSCURE_CINT_HACK

typedef list<char *> cplist;

cplist filenames;

void cleanup(void)
{
    for (cplist::iterator it=filenames.begin(); it!=filenames.end(); ++it) {
        cout << "removing file " << *it << endl;
        remove(*it);
        free(*it);
    }
}

void sigterm_handler(int n)
{
    exit(1);
}

bool is_whitespace_char(char c)
{
    return ((c==' ') || (c=='\t') || (c=='\n') || (c=='\r')); 
}

char first_non_whitespace_char(char const * s)
{
    int l = strlen(s);
    int pos = 0;
    while ((pos<l)&&is_whitespace_char(s[pos])) pos++;
    return s[pos];
}    

char last_non_whitespace_char(char const * s)
{
    int pos = strlen(s)-1;
    while ((pos>=0) && is_whitespace_char(s[pos])) pos--;
    return s[pos];
}    

string strip_whitespace(string const & s)
{
    string s2;
    int l = s.length();
    for (int pos=0; pos<l; ++pos) {
        if (!is_whitespace_char(s[pos])) s2 += s[pos];
    }
    return s2;
}

G__value exec_tempfile(string const & command)
{
    G__value retval;
    char *tmpfilename = tempnam(NULL,"ginac");
    ofstream fout;
    fout.open(tmpfilename);
    fout << "{" << endl << command << endl << "}" << endl;
    fout.close();
    G__store_undo_position();
    retval=G__exec_tempfile(tmpfilename);
    G__security_recover(stdout);
    remove(tmpfilename);
    free(tmpfilename);
    return retval;
}

char * process_permanentfile(string const & command)
{
    char *tmpfilename = tempnam(NULL,"ginac");
    cout << "creating file " << tmpfilename << endl;
    ofstream fout;
    fout.open(tmpfilename);
    fout << command << endl;
    fout.close();
    G__store_undo_position();
    G__loadfile(tmpfilename);
    G__security_recover(stdout);
    return tmpfilename;
}

void process_tempfile(string const & command)
{
#ifdef OBSCURE_CINT_HACK
    static G__value ref_symbol = exec_tempfile("symbol ginac_cint_internal_symbol; ginac_cint_internal_symbol;");
    static G__value ref_ex = exec_tempfile("ex ginac_cint_internal_ex; ginac_cint_internal_ex;");
    static G__value ref_function = exec_tempfile("sin(ginac_cint_internal_symbol);");
    static G__value ref_power = exec_tempfile("power(ginac_cint_internal_symbol,ginac_cint_internal_symbol);");
#endif // def OBSCURE_CINT_HACK

    G__value retval = exec_tempfile(command);

#ifdef OBSCURE_CINT_HACK

    #define TYPES_EQUAL(A,B) (((A).type==(B).type) && ((A).tagnum==(B).tagnum))
    
    static unsigned out_count = 0;
    if (TYPES_EQUAL(retval,ref_ex)) {
        string varname = "Out"+ToString(++out_count);
        if (retval.obj.i!=ex::last_created_or_assigned_exp) {
            // an ex was returned, but this is not the ex which was created last
            // => this is not a temporary ex, but one that resides safely in memory
            
            // cout << "warning: using ex from retval (experimental)" << endl;
            ex::dummy_bp=((ex *)(void *)(retval.obj.i))->bp;
            exec_tempfile("ex "+varname+"(*ex::dummy_bp);");
        } else if (ex::last_created_or_assigned_bp_can_be_converted_to_ex()) {
            //string varfill;
            //for (int i=4-int(log10(out_count)); i>0; --i)
            //    varfill += ' ';
            exec_tempfile("ex "+varname+"(*ex::last_created_or_assigned_bp);");
        } else {
            cout << "warning: last_created_or_assigned_bp modified 0 or not evaluated or not dynallocated" << endl;
        }
        exec_tempfile(string()+"LLLAST=LLAST;\n"
                      +"LLAST=LAST;\n"
                      +"LAST="+varname+";\n"
                      +"cout << \""+varname+" = \" << "+varname+" << endl << endl;");
    }
#endif // def OBSCURE_CINT_HACK
}

void greeting(void)
{
    cout << "Welcome to GiNaC-cint (" << PACKAGE << " V" << VERSION << ")" << endl;
    cout << "This software is provided \"as is\" without any warranty.  Copyright of Cint is" << endl
         << "owned by Agilent Technologies Japan and Masaharu Goto.  Registration is" << endl
         << "  __,  _______  requested, at this moment, for commercial use.  Send e-mail to" << endl
         << " (__) *       | <MXJ02154@niftyserve.or.jp>.  The registration is free." << endl
         << "  ._) i N a C | The GiNaC framework is Copyright by Johannes Gutenberg Univ.," << endl
         << "<-------------' Germany and licensed under the terms and conditions of the GPL." << endl << endl;
}

int main(void) 
{
    char *line;
    char prompt[G__ONELINE];

    greeting();

    atexit(cleanup);
    signal(SIGTERM,sigterm_handler);
    
    G__init_cint("cint");    /* initialize cint */

    exec_tempfile("#include <string>\n");
    exec_tempfile("ex LAST,LLAST,LLLAST;\n");
    
    bool quit = false;
    bool next_command_is_function=false;    
    while (!quit) {
        strcpy(prompt,"GiNaC> ");
        int open_braces = 0;
        bool end_of_command=false;
        string command;
        while (!end_of_command) {
            line = G__input(prompt);
        
            int pos = 0;
            bool double_quote=false;
            bool single_quote=false;
            while(line[pos]!='\0') {
                switch(line[pos]) {
                case '"':
                    if (!single_quote) double_quote = !double_quote;
                    break;
                case '\'':
                    if (!double_quote) single_quote = !single_quote;
                    break;
                case '{':
                    if ((!single_quote) && (!double_quote)) open_braces++;
                    break;
                case '}':
                    if ((!single_quote) && (!double_quote)) open_braces--;
                    break;
                }
                pos++;
            }
            command += line;
            command += '\n';
            if (open_braces==0) {
                if ((first_non_whitespace_char(command.c_str())=='#')||
                    (first_non_whitespace_char(command.c_str())=='.')||
                    (last_non_whitespace_char(command.c_str())==';')||
                    (last_non_whitespace_char(command.c_str())=='}')) {
                    end_of_command=true;
                }
            }
            strcpy(prompt,"     > ");
        }
        string stripped_command=strip_whitespace(command);
        if ((stripped_command=="quit;")||
            (stripped_command=="exit;")||
            (stripped_command=="bye;")||
            (stripped_command==".q")||
            (stripped_command==".quit")||
            (stripped_command==".exit")||
            (stripped_command==".bye")) {
            quit = true;
        } else if (stripped_command==".function") {
            cout << "next expression can be a function definition" << endl;
            next_command_is_function=true;
        } else if (stripped_command==".cint") {
            cout << endl << "switching to cint interactive mode" << endl;
            cout << "'h' for help, 'q' to quit, '{ statements }' or 'p [expression]' to evaluate" << endl;
            G__pause();
            cout << "back from cint" << endl;
        } else if (command[0]=='.') {
            cout << "special command (TBD): " << command << endl;
        } else {
            // cout << "now processing: " << command << endl;
            if (next_command_is_function) {
                next_command_is_function = false;
                filenames.push_back(process_permanentfile(command));
            } else {
                process_tempfile(command);
            }
        }
    }

    return 0;
}







